ENH: Add numba engine to df.apply

lithomas1 opened the pull request:
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

~~Performance is a mixed bag, it is sometimes faster and sometimes slower than Python.
This is because there is a huge cost (esp. in boxing the numba representations back into Series/Index since the constructors are slow). 
Right now, in the no-op case(return self), we are around 5-10x slower - I'm hoping to bring that down to around 2x.
(There is an issue with the Index being repeatedly unboxed unnecessarily)~~

EDIT: I think I've closed the gap, the difference isn't much now. 

We are also ~10x faster in a normalization (subtract mean, divide by std. dev) test I did.

lithomas1 committed:
ENH: Add numba engine to df.apply

lithomas1 committed:
Merge branch 'main' of github.com:pandas-dev/pandas into numba-apply

lithomas1 committed:
complete?

lithomas1 committed:
wip: pass tests

lithomas1 committed:
Merge branch 'main' of github.com:pandas-dev/pandas into numba-apply

lithomas1 committed:
fix existing tests

lithomas1 committed:
go for green

lithomas1 committed:
fix checks?

lithomas1 committed:
fix pyright

lithomas1 commented:
@mroeschke 

This should be ready for a first pass now.
I know this is a lot of code, as of right now, this PR implements a pretty minimal subset of the pandas features, which are:

1. Numba versions of DataFrame, Series, Index
2. Common Series methods that have a numpy equivalent
   a. Think sum, var, mean, and co.
3. Basic indexing support
   a. It mimics our khash stuff, just with a numba dictionary instead. We might look into re-using the khash code, it would require some Cython code to expose that to numba though.

As of right now, there isn't support for things such as the Arrow dtypes, and non-numeric Indexes, but those shouldn't be hard to add in the future - I just didn't think the maintenance burden would be worth it as of right now.

(If you want, I can try to split up the PR, but because 1 is still a lot of lines, its still gonna be pretty big to review unfortunately).

Performance is decent as of now.
We are generally anywhere between 2-10x faster when going row-wise, but can be slower in cases where there are few rows/columns.

Most of the time is spent boxing/unboxing (converting to/from) the numba representation, and the pandas representation of the DataFrame. In the future, we could optimize this to potentially get like 100x speedups.

This should be fairly fixable with some effort, when just need to re-implement the concat/results wrapping that apply does in numba. I've held off on this because I've written a lot of code so far, and like before, I'd like to see people use this feature before I add to the maintenance burden of the numba code.

Sadly, multithreading doesn't yet work - numba has no thread safe structures, but this won't be relevant until the unboxing/boxing overhead can be dealt with.

lithomas1 commented on pandas/core/_numba/extensions.py:
```diff
@@ -0,0 +1,541 @@
+
+# TODO: Range index support
+# (not passing an index to series constructor doesn't work)
+class IndexType(types.Type):
```
This block just defines the types for Index and Series, there isn't much to see here.

It is pretty boilerplate and standard.

lithomas1 commented on pandas/core/_numba/extensions.py:
```diff
@@ -0,0 +1,541 @@
+
+
+@typeof_impl.register(Index)
+def typeof_index(val, c):
```
This block is where we teach numba to recognize our pandas objects as something that can be lowered into numba.

Again, pretty boilerplate and standard.

lithomas1 commented on pandas/core/_numba/extensions.py:
```diff
@@ -0,0 +1,541 @@
+
+
+@type_callable(Series)
+def type_series_constructor(context):
```
These define the types that the Series/Index constructors take in when you call the Series/Index constructors inside numba code.

It's fairly uninteresting. Note that the actual implementation of the constructors are further down - think of these declarations as kind of like a C prototype function.

lithomas1 commented on pandas/core/_numba/extensions.py:
```diff
@@ -0,0 +1,541 @@
+
+
+# Backend extensions for Index and Series and Frame
+@register_model(IndexType)
```
This defines the numba representations of index/series.

Only interesting thing here is that Index has a pointer to the original index object, so we can avoid calling the index constructor and then just return that object.

Also, we add a hashmap to the index for indexing purposes.

mroeschke replied:
Does the hashmap support duplicate values like a pandas Index would?

lithomas1 replied:
Good point, I will update and add some tests.

lithomas1 replied:
On second thought, it's probably easier for now to disallow duplicate indexes.

I don't most frames have duplicate columns/indexes.

lithomas1 commented on pandas/core/_numba/extensions.py:
```diff
@@ -0,0 +1,541 @@
+
+
+@lower_builtin(Series, types.Array, IndexType)
+def pdseries_constructor(context, builder, sig, args):
```
Constructor implementations.

lithomas1 commented on pandas/core/_numba/extensions.py:
```diff
@@ -0,0 +1,541 @@
+
+
+@unbox(IndexType)
+def unbox_index(typ, obj, c):
```
Code that transforms Series/Index -> numba representations and back.

There's a lot of C API stuff here, it's maybe worth a closer look if you want.

lithomas1 commented on pandas/core/_numba/extensions.py:
```diff
@@ -0,0 +1,541 @@
+# and also add common binops (e.g. add, sub, mul, div)
+
+
+def generate_series_reduction(ser_reduction, ser_method):
```
Code to generate reductions (mean, std, var) and binops (addition, subtraction, multiplication)

lithomas1 commented on pandas/core/_numba/extensions.py:
```diff
@@ -0,0 +1,541 @@
+
+
+# get_loc on Index
+@overload_method(IndexType, "get_loc")
```
The indexing code goes from here to the end of the file, its maybe worth a closer look if you want to make sure it aligns with what we actually do.

lithomas1 reviewed (commented):
Added some comments to make review easier.

lithomas1 committed:
update docs

lithomas1 committed:
Merge branch 'main' of github.com:pandas-dev/pandas into numba-apply

lithomas1 committed:
Merge branch 'main' into numba-apply

lithomas1 committed:
eliminate a blank line

lithomas1 commented:
gentle ping @mroeschke

mroeschke commented on pandas/core/apply.py:
```diff
@@ -1060,6 +1117,56 @@ class FrameRowApply(FrameApply):
+                )
+            col_names_values = orig_values.astype("U")
+            # Remember to set this back!
+            self.columns._data = col_names_values
```
Why does this need to be assigned to `._data`? Couldn't a copy of the columns just be passed to `nb_func`?

mroeschke replied:
Oh is this needed due to how the numba extension is defined?

lithomas1 replied:
Yeah, in Index, we don't allow numpy string dtypes, but my hack uses numpy string dtypes, since those already have a native representation in numba.

Let me know if this is too hacky.

mroeschke commented on pandas/core/apply.py:
```diff
@@ -1060,6 +1117,56 @@ class FrameRowApply(FrameApply):
+        @numba.jit(nogil=nogil, nopython=nopython, parallel=parallel)
+        def numba_func(values, col_names, df_index):
+            results = {}
+            for j in range(values.shape[1]):
```
```suggestion
            for j in numba.prange(values.shape[1]):
```

? (and below)

lithomas1 replied:
Thanks for pointing this out!

I think for now it'll probably make better sense to disable parallel mode for now, since the dict in numba isn't thread-safe.

The overhead from the boxing/unboxing is also really high (99% of the time spent is there), so I doubt parallel will give a good speedup, at least for now.

mroeschke replied:
OK makes sense. Would be good to put a TODO: comment explaining why we shouldn't use prange for now

lithomas1 replied:
added a comment.

mroeschke commented on pandas/core/_numba/extensions.py:
```diff
@@ -0,0 +1,539 @@
+import numba
+from numba.core import (
+    cgutils,
+    types,
```
Could we just access this from `numba.types`?

lithomas1 commented on pandas/core/apply.py:
```diff
@@ -1060,6 +1117,56 @@ class FrameRowApply(FrameApply):
+            results = {}
+            for j in range(values.shape[1]):
+                # Create the series
+                ser = Series(values[:, j], index=df_index, name=str(col_names[j]))
```
I restrict it to only allow string names.
(e.g. here https://github.com/pandas-dev/pandas/pull/55104/files#diff-2257b34410aee27eb14e348b9545fef2e212ff93bd72af02d700ae8df43d97bbR1153-R1158)

The cast to string is a quirk of my hack.
(I convert the column names to a numpy string array, so each element is a numpy string that I then need to convert to a regular numba unicode value)

lithomas1 replied:
Nvm, I think I see what you mean.

I think I need to change it so it only cassts to string when the index is str dtype.

lithomas1 committed:
update from code review + more tests

lithomas1 committed:
Merge branch 'main' of github.com:pandas-dev/pandas into numba-apply

lithomas1 committed:
fix failing tests

lithomas1 committed:
Simplify w/ context manager

lithomas1 committed:
skip if no numba

lithomas1 committed:
simplify more

lithomas1 commented:
@mroeschke 
Can you take another look at this?

I've simplified the hacky parts a bit so we no longer clobber the ``_data`` attribute of an Index, and added some more tests.

There's still some work to be done on catching and testing all the unsupported cases(e.g. EAs, Arrow arrays, object dtype, etc.), but I'll put that in a followup PR to keep this one small.

lithomas1 committed:
specify dtypes

lithomas1 committed:
Merge branch 'main' of github.com:pandas-dev/pandas into numba-apply

mroeschke commented on pandas/tests/apply/test_numba.py:
```diff
@@ -0,0 +1,64 @@
+
+def test_numba_nonunique_unsupported():
+    f = lambda x: x
+    df = DataFrame({"a": [1, 2], "b": [1, 2]})
```
Doesn't this have unique index and columns?

lithomas1 replied:
You're on an old commit?

mroeschke commented on pandas/tests/apply/test_numba.py:
```diff
@@ -0,0 +1,64 @@
+    expected = float_frame.apply(row_func, engine="python", axis=1)
+    tm.assert_series_equal(result, expected)
+
+    row_func = lambda x: x["ZqgszYBfuL"]  # This is a label in the index
```
Nit: I think it would be okay using a different fixture (specific to this file) than `float_frame`. IMO it can be a little more work to go back and debug not knowing what `float_frame` exactly is

lithomas1 replied:
Sure, will do.

mroeschke commented on pandas/core/_numba/extensions.py:
```diff
@@ -0,0 +1,575 @@
+# that is the original data, or the object data casted to numpy string dtype,
+# with a context manager that is unset afterwards
+@contextmanager
+def set_numba_data(index: Index):
```
I think you might be able to reuse `common.temp_setattr` for this

mroeschke replied:
Ok sounds good

lithomas1 committed:
Merge branch 'main' into numba-apply

lithomas1 committed:
Merge branch 'numba-apply' of github.com:lithomas1/pandas into numba-apply

lithomas1 committed:
address code review

mroeschke commented:
Final question. What error do we currently get when trying to use apply with numba with an unsupported type (like `datetime64[ns]`?

lithomas1 committed:
add errors for invalid columns

lithomas1 commented:
> Final question. What error do we currently get when trying to use apply with numba with an unsupported type (like `datetime64[ns]`?

I added an error message and a test.

It's basically a ValueError saying the dtype in this column is unsupported.

lithomas1 committed:
adjust message

mroeschke commented:
Great work on this @lithomas1. In a follow up, could you add a whatsnew note in 2.2?

lithomas1 commented:
Thanks for the reviews - I think there should be a whatnsew from when I added support with raw apply.

This definetely deserves a bigger blurb, though. I'll add that sometime next week.
